File: identity/src/main/java/com/android/identity/GattServer.java
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattServer;
import android.bluetooth.BluetoothGattServerCallback;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.ByteArrayOutputStream;
import java.util.ArrayDeque;
import java.util.Locale;
import java.util.Queue;
import java.util.UUID;

@SuppressWarnings("deprecation")
@SuppressLint("MissingPermission")
class GattServer extends BluetoothGattServerCallback {
    private static final String TAG = "GattServer";
    private final byte[] mEncodedEDeviceKeyBytes;
    private final UUID mServiceUuid;
    private final BluetoothManager mBluetoothManager;
    private final Context mContext;
    UUID mCharacteristicStateUuid;
    UUID mCharacteristicClient2ServerUuid;
    UUID mCharacteristicServer2ClientUuid;
    UUID mCharacteristicIdentUuid;
    UUID mCharacteristicL2CAPUuid;

    // This is what the 16-bit UUID 0x29 0x02 is encoded like.
    UUID mClientCharacteristicConfigUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");
    Listener mListener;
    boolean mInhibitCallbacks = false;
    BluetoothGattCharacteristic mCharacteristicState;
    BluetoothGattCharacteristic mCharacteristicClient2Server;
    BluetoothGattCharacteristic mCharacteristicServer2Client;
    BluetoothGattCharacteristic mCharacteristicIdent;
    BluetoothGattCharacteristic mCharacteristicL2CAP;
    ByteArrayOutputStream mIncomingMessage = new ByteArrayOutputStream();
    Queue<byte[]> mWritingQueue = new ArrayDeque<>();
    int writingQueueTotalChunks;
    boolean writeIsOutstanding = false;
    private BluetoothGattServer mGattServer;
    private BluetoothDevice mCurrentConnection;
    private int mNegotiatedMtu = 0;
    private L2CAPServer mL2CAPServer;
    private byte[] mIdentValue;
    private boolean mUsingL2CAP = false;

    GattServer(@NonNull Context context,
               @NonNull BluetoothManager bluetoothManager,
               @NonNull UUID serviceUuid,
               @Nullable byte[] encodedEDeviceKeyBytes,
               @NonNull UUID characteristicStateUuid,
               @NonNull UUID characteristicClient2ServerUuid,
               @NonNull UUID characteristicServer2ClientUuid,
               @Nullable UUID characteristicIdentUuid,
               @Nullable UUID characteristicL2CAPUuid) {
        mContext = context;
        mBluetoothManager = bluetoothManager;
        mServiceUuid = serviceUuid;
        mEncodedEDeviceKeyBytes = encodedEDeviceKeyBytes;
        mCharacteristicStateUuid = characteristicStateUuid;
        mCharacteristicClient2ServerUuid = characteristicClient2ServerUuid;
        mCharacteristicServer2ClientUuid = characteristicServer2ClientUuid;
        mCharacteristicIdentUuid = characteristicIdentUuid;
        mCharacteristicL2CAPUuid = characteristicL2CAPUuid;
    }

    void setListener(@Nullable Listener listener) {
        mListener = listener;
    }

    @SuppressLint("NewApi")
    boolean start() {
        if (mEncodedEDeviceKeyBytes != null) {
            byte[] ikm = mEncodedEDeviceKeyBytes;
            byte[] info = new byte[]{'B', 'L', 'E', 'I', 'd', 'e', 'n', 't'};
            byte[] salt = new byte[]{};
            mIdentValue = Util.computeHkdf("HmacSha256", ikm, salt, info, 16);
        }

        try {
            mGattServer = mBluetoothManager.openGattServer(mContext, this);
        } catch (SecurityException e) {
            reportError(e);
            return false;
        }
        if (mGattServer == null) {
            return false;
        }

        BluetoothGattService service = new BluetoothGattService(mServiceUuid,
                BluetoothGattService.SERVICE_TYPE_PRIMARY);

        BluetoothGattCharacteristic c;
        BluetoothGattDescriptor d;

        // State
        c = new BluetoothGattCharacteristic(mCharacteristicStateUuid,
                BluetoothGattCharacteristic.PROPERTY_NOTIFY
                        | BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        d = new BluetoothGattDescriptor(mClientCharacteristicConfigUuid,
                BluetoothGattDescriptor.PERMISSION_WRITE);
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE);
        c.addDescriptor(d);
        service.addCharacteristic(c);
        mCharacteristicState = c;

        // Client2Server
        c = new BluetoothGattCharacteristic(mCharacteristicClient2ServerUuid,
                BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        service.addCharacteristic(c);
        mCharacteristicClient2Server = c;

        // Server2Client
        c = new BluetoothGattCharacteristic(mCharacteristicServer2ClientUuid,
                BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_WRITE);
        d = new BluetoothGattDescriptor(mClientCharacteristicConfigUuid,
                BluetoothGattDescriptor.PERMISSION_WRITE);
        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE);
        c.addDescriptor(d);
        service.addCharacteristic(c);
        mCharacteristicServer2Client = c;

        // Ident
        if (mCharacteristicIdentUuid != null) {
            c = new BluetoothGattCharacteristic(mCharacteristicIdentUuid,
                    BluetoothGattCharacteristic.PROPERTY_READ,
                    BluetoothGattCharacteristic.PERMISSION_READ);
            service.addCharacteristic(c);
            mCharacteristicIdent = c;
        }

        // Offers support to L2CAP when we have UUID characteristic and the OS version support it
        mUsingL2CAP = (mCharacteristicL2CAPUuid != null) && (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q);         
        Logger.d(TAG, "Is L2CAP supported: " + mUsingL2CAP);
        if (mUsingL2CAP) {
            c = new BluetoothGattCharacteristic(mCharacteristicL2CAPUuid,
                    BluetoothGattCharacteristic.PROPERTY_READ,
                    BluetoothGattCharacteristic.PERMISSION_READ);
            service.addCharacteristic(c);
            mCharacteristicL2CAP = c;
        }

        try {
            mGattServer.addService(service);
        } catch (SecurityException e) {
            reportError(e);
            return false;
        }

        return true;
    }

    void stop() {
        mInhibitCallbacks = true;
        if (mL2CAPServer != null) {
            mL2CAPServer.stop();
        }
        if (mGattServer != null) {
            try {
                if (mCurrentConnection != null) {
                    mGattServer.cancelConnection(mCurrentConnection);
                }
                mGattServer.close();
            } catch (SecurityException e) {
                Logger.e(TAG, "Caught SecurityException while shutting down: " + e);
            }
        }
    }

    @Override
    public void onConnectionStateChange(BluetoothDevice device, int status, int newState) {
        Logger.d(TAG, "onConnectionStateChange: " + device.getAddress() + " " + status + " " + newState);
        if (newState == BluetoothProfile.STATE_DISCONNECTED
                && mCurrentConnection != null
                && device.getAddress().equals(mCurrentConnection.getAddress())) {
            Logger.d(TAG, "Device " + mCurrentConnection.getAddress() + " which we're currently "
                    + "connected to, has disconnected");
            mCurrentConnection = null;
            reportPeerDisconnected();
        }
    }

    @SuppressLint("NewApi")
    @Override
    public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, int offset,
                                            BluetoothGattCharacteristic characteristic) {
        Logger.d(TAG, "onCharacteristicReadRequest: "  + device.getAddress() + " "
                + requestId + " " + offset + " " + characteristic.getUuid());
        if (mCharacteristicIdentUuid != null
                && characteristic.getUuid().equals(mCharacteristicIdentUuid)) {
            try {
                byte[] ident = mIdentValue != null ? mIdentValue : new byte[0];
                mGattServer.sendResponse(device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        0,
                        ident);
            } catch (SecurityException e) {
                reportError(e);
            }
        } else if (mCharacteristicL2CAP != null
                && characteristic.getUuid().equals(mCharacteristicL2CAPUuid)) {
            if (!mUsingL2CAP) {
                reportError(new Error("Unexpected read request for L2CAP characteristic, not supported"));
                return;
            }

            // Start L2CAP socket server
            mL2CAPServer = new L2CAPServer(new L2CAPServer.Listener() {
                @Override
                public void onPeerConnected() {
                    reportPeerConnected();
                }

                @Override
                public void onPeerDisconnected() {
                    reportPeerDisconnected();
                }

                @Override
                public void onMessageReceived(@NonNull byte[] data) {
                    reportMessageReceived(data);
                }

                @Override
                public void onError(@NonNull Throwable error) {
                    reportError(error);
                }

                @Override
                public void onMessageSendProgress(long progress, long max) {
                    reportMessageSendProgress(progress, max);
                }
            });

            byte[] psmValue = mL2CAPServer.start(mBluetoothManager.getAdapter());
            if (psmValue == null) {
                mUsingL2CAP = false;
            } else {
                mGattServer.sendResponse(device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        0,
                        psmValue);
            }
        } else {
            reportError(new Error("Read on unexpected characteristic with UUID "
                    + characteristic.getUuid()));
        }
    }

    @Override
    public void onCharacteristicWriteRequest(BluetoothDevice device,
                                             int requestId,
                                             BluetoothGattCharacteristic characteristic,
                                             boolean preparedWrite,
                                             boolean responseNeeded,
                                             int offset,
                                             byte[] value) {
        UUID charUuid = characteristic.getUuid();
        Logger.d(TAG, "onCharacteristicWriteRequest: " + device.getAddress() + " "
                + characteristic.getUuid() + " " + offset + " " + Util.toHex(value));

        // If we are connected to a device, ignore write from any other device
        if (mCurrentConnection != null &&
                !device.getAddress().equals(mCurrentConnection.getAddress())) {
            Logger.e(TAG, "Ignoring characteristic write request from "
            + device.getAddress() + " since we're already connected to "
            + mCurrentConnection.getAddress());
            return;
        }

        if (charUuid.equals(mCharacteristicStateUuid) && value.length == 1) {
            if (value[0] == 0x01) {
                // Close server socket when the connection was done by state characteristic
                stopL2CAPServer();
                if (mCurrentConnection != null) {
                    Logger.e(TAG, "Ignoring connection attempt from " + device.getAddress()
                    + " since we're already connected to " + mCurrentConnection.getAddress());
                } else {
                    mCurrentConnection = device;
                    Logger.d(TAG, "Received connection (state 0x01 on State characteristic) "
                            + "from " + mCurrentConnection.getAddress());
                }
                reportPeerConnected();
            } else if (value[0] == 0x02) {
                reportTransportSpecificSessionTermination();
            } else {
                reportError(new Error("Invalid byte " + value[0] + " for state "
                        + "characteristic"));
            }
        } else if (charUuid.equals(mCharacteristicClient2ServerUuid)) {
            if (value.length < 1) {
                reportError(new Error("Invalid value with length " + value.length));
                return;
            }
            if (mCurrentConnection == null) {
                // We expect a write 0x01 on the State characteristic before we consider
                // the device to be connected.
                reportError(new Error("Write on Client2Server but not connected yet"));
                return;
            }
            mIncomingMessage.write(value, 1, value.length - 1);
            Logger.d(TAG, String.format(Locale.US,
                    "Received chunk with %d bytes (last=%s), incomingMessage.length=%d",
                    value.length, value[0] == 0x00, mIncomingMessage.toByteArray().length));
            if (value[0] == 0x00) {
                // Last message.
                byte[] entireMessage = mIncomingMessage.toByteArray();
                mIncomingMessage.reset();
                reportMessageReceived(entireMessage);
            } else if (value[0] == 0x01) {
                if (value.length != getCharacteristicValueSize()) {
                    Logger.w(TAG, String.format(Locale.US,
                            "Client2Server received %d bytes which is not the expected %d bytes",
                            value.length, getCharacteristicValueSize()));
                    return;
                }
            } else {
                reportError(new Error(String.format(Locale.US,
                        "Invalid first byte %d in Client2Server data chunk, expected 0 or 1",
                        value[0])));
                return;
            }

            if (responseNeeded) {
                try {
                    mGattServer.sendResponse(device,
                            requestId,
                            BluetoothGatt.GATT_SUCCESS,
                            0,
                            null);
                } catch (SecurityException e) {
                    reportError(e);
                }
            }
        } else {
            reportError(new Error("Write on unexpected characteristic with UUID "
                    + characteristic.getUuid()));
        }
    }

    private void stopL2CAPServer() {
        // Close server socket when the connection was done by state characteristic
        if (mL2CAPServer != null) {
            mL2CAPServer.stop();
            mL2CAPServer = null;
        }
        mUsingL2CAP = false;
    }

    @Override
    public void onDescriptorReadRequest(BluetoothDevice device, int requestId, int offset,
                                        BluetoothGattDescriptor descriptor) {
        Logger.d(TAG, "onDescriptorReadRequest: " + device.getAddress() + " "
                + descriptor.getCharacteristic().getUuid() + " " + offset);
        /* Do nothing */
    }

    @Override
    public void onDescriptorWriteRequest(BluetoothDevice device, int requestId,
                                         BluetoothGattDescriptor descriptor,
                                         boolean preparedWrite, boolean responseNeeded,
                                         int offset, byte[] value) {
        if (Logger.isDebugEnabled()) {
            Logger.d(TAG, "onDescriptorWriteRequest: " + device.getAddress() + " "
                    + descriptor.getCharacteristic().getUuid() + " " + offset + " "
                    + Util.toHex(value));
        }
        if (responseNeeded) {
            try {
                mGattServer.sendResponse(device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        0,
                        null);
            } catch (SecurityException e) {
                reportError(e);
            }
        }
    }

    @Override
    public void onMtuChanged(BluetoothDevice device, int mtu) {
        mNegotiatedMtu = mtu;
        Logger.d(TAG, "Negotiated MTU " + mtu + " for " + device.getAddress() + " ");
    }

    private int mCharacteristicValueSize = 0;

    private int
    getCharacteristicValueSize() {
        if (mCharacteristicValueSize > 0) {
            return mCharacteristicValueSize;
        }
        int mtuSize = mNegotiatedMtu;
        if (mtuSize == 0) {
            Logger.w(TAG, "MTU not negotiated, defaulting to 23. Performance will suffer.");
            mtuSize = 23;
        }
        mCharacteristicValueSize = Util.bleCalculateAttributeValueSize(mtuSize);
        return mCharacteristicValueSize;
    }

    void drainWritingQueue() {
        Logger.d(TAG, "drainWritingQueue " + writeIsOutstanding);
        if (writeIsOutstanding) {
            return;
        }
        byte[] chunk = mWritingQueue.poll();
        if (chunk == null) {
            return;
        }

        Logger.d(TAG, String.format(Locale.US,
                "Sending chunk with %d bytes (last=%s)",
                chunk.length, chunk[0] == 0x00));
        mCharacteristicServer2Client.setValue(chunk);
        try {
            if (!mGattServer.notifyCharacteristicChanged(mCurrentConnection,
                    mCharacteristicServer2Client, false)) {
                reportError(
                        new Error("Error calling notifyCharacteristicsChanged on Server2Client"));
                return;
            }
        } catch (SecurityException e) {
            reportError(e);
            return;
        }
        writeIsOutstanding = true;
    }

    @Override
    public void onNotificationSent(BluetoothDevice device, int status) {
        Logger.d(TAG, "onNotificationSent " + status + " for " + device.getAddress() + " ");
        if (status != BluetoothGatt.GATT_SUCCESS) {
            reportError(new Error("Error in onNotificationSent status=" + status));
            return;
        }

        if (writingQueueTotalChunks > 0) {
            if (mWritingQueue.size() == 0) {
                reportMessageSendProgress(writingQueueTotalChunks, writingQueueTotalChunks);
                writingQueueTotalChunks = 0;
            } else {
                reportMessageSendProgress(writingQueueTotalChunks - mWritingQueue.size(), writingQueueTotalChunks);
            }
        }


        writeIsOutstanding = false;
        drainWritingQueue();
    }

    void sendMessage(@NonNull byte[] data) {
        Logger.dHex(TAG, "sendMessage", data);

        // Uses socket L2CAP when it is available
        if (mL2CAPServer != null) {
            mL2CAPServer.sendMessage(data);
            return;
        }

        // Also need room for the leading 0x00 or 0x01.
        //
        int maxDataSize = getCharacteristicValueSize() - 1;

        int offset = 0;
        do {
            boolean moreDataComing = (offset + maxDataSize < data.length);
            int size = data.length - offset;
            if (size > maxDataSize) {
                size = maxDataSize;
            }

            byte[] chunk = new byte[size + 1];
            chunk[0] = moreDataComing ? (byte) 0x01 : (byte) 0x00;
            System.arraycopy(data, offset, chunk, 1, size);

            mWritingQueue.add(chunk);

            offset += size;
        } while (offset < data.length);
        writingQueueTotalChunks = mWritingQueue.size();
        drainWritingQueue();
    }

    void reportPeerConnected() {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onPeerConnected();
        }
    }

    void reportPeerDisconnected() {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onPeerDisconnected();
        }
    }

    void reportMessageReceived(@NonNull byte[] data) {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onMessageReceived(data);
        }
    }

    void reportMessageSendProgress(long progress, long max) {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onMessageSendProgress(progress, max);
        }
    }

    void reportError(@NonNull Throwable error) {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onError(error);
        }
    }

    void reportTransportSpecificSessionTermination() {
        if (mListener != null && !mInhibitCallbacks) {
            mListener.onTransportSpecificSessionTermination();
        }
    }

    // When using L2CAP it doesn't support characteristics notification
    public boolean supportsTransportSpecificTerminationMessage() {
        return !mUsingL2CAP;
    }

    public void sendTransportSpecificTermination() {
        byte[] terminationCode = new byte[]{(byte) 0x02};
        mCharacteristicState.setValue(terminationCode);
        try {
            if (!mGattServer.notifyCharacteristicChanged(mCurrentConnection,
                    mCharacteristicState, false)) {
                reportError(new Error("Error calling notifyCharacteristicsChanged on State"));
            }
        } catch (SecurityException e) {
            reportError(e);
        }
    }

    interface Listener {
        void onPeerConnected();

        void onPeerDisconnected();

        void onMessageReceived(@NonNull byte[] data);

        void onTransportSpecificSessionTermination();

        void onError(@NonNull Throwable error);

        void onMessageSendProgress(long progress, long max);

    }
}

File: bleofflinetransfer/src/main/java/com/ul/ims/gmdl/bleofflinetransfer/peripheral/GattServer.kt
/*
 * Copyright (C) 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

package com.ul.ims.gmdl.bleofflinetransfer.peripheral

import android.bluetooth.*
import android.bluetooth.BluetoothGattDescriptor.PERMISSION_WRITE
import android.content.Context
import android.util.Log
import com.ul.ims.gmdl.bleofflinetransfer.*
import com.ul.ims.gmdl.bleofflinetransfer.common.BleEventListener
import com.ul.ims.gmdl.bleofflinetransfer.common.DataNotifier
import com.ul.ims.gmdl.bleofflinetransfer.common.GattCommon
import com.ul.ims.gmdl.bleofflinetransfer.config.ServiceCharacteristics
import com.ul.ims.gmdl.bleofflinetransfer.exceptions.BluetoothException
import com.ul.ims.gmdl.bleofflinetransfer.exceptions.GattException
import com.ul.ims.gmdl.bleofflinetransfer.exceptions.TransportLayerException
import com.ul.ims.gmdl.bleofflinetransfer.utils.ByteUtils
import com.ul.ims.gmdl.bleofflinetransfer.utils.ByteUtils.decodeData
import com.ul.ims.gmdl.bleofflinetransfer.utils.ByteUtils.deviceString
import com.ul.ims.gmdl.offlinetransfer.executorLayer.IExecutorEventListener
import com.ul.ims.gmdl.offlinetransfer.transportLayer.EventType
import java.io.ByteArrayOutputStream
import java.util.*

class GattServer (
    private val context: Context,
    private val bluetoothManager: BluetoothManager,
    private val bleServiceCharacteristics: ServiceCharacteristics,
    private val publicKeyHash: ByteArray,
    override var bleEventListener: BleEventListener?
) : GattCommon, BluetoothGattServerCallback() {

    override var stateCharacteristic: BluetoothGattCharacteristic? = null
    override var serverToClientCharacteristic: BluetoothGattCharacteristic? = null
    override var clientToServerCharacteristic: BluetoothGattCharacteristic? = null
    override var identCharacteristic: BluetoothGattCharacteristic? = null
    override var maxSupportedMtu: Int = MAX_MTU
    override var dataStream: ByteArrayOutputStream = ByteArrayOutputStream()
    override var executorEventListener: IExecutorEventListener? = null
    override var bluetoothGattService: BluetoothGattService? = null

    private var readyForNextFile: Boolean = true
    private var bluetoothGattServer: BluetoothGattServer? = null
    private var currentDevice: BluetoothDevice? = null
    private var requestId: Int = 0
    private var offset: Int = 0
    private var dataNotifier: DataNotifier? = null

    init {
        setup()
    }

    companion object {
        val LOG_TAG = "GattServer"
    }

    /**
     * Setup the Characteristics to be advertised
     * **/
    private fun setup() {
        try {

            bluetoothGattServer = bluetoothManager.openGattServer(context, this)
            bluetoothGattServer?.let {

                bluetoothGattService = BluetoothGattService(bleServiceCharacteristics.serviceUuid,
                    BluetoothGattService.SERVICE_TYPE_PRIMARY)

                // Server2Client
                serverToClientCharacteristic = bleServiceCharacteristics.server2ClientCharacteristic
                val descriptorServerToClient = BluetoothGattDescriptor(
                    UUID.fromString(CLIENT_CHARACTERISTIC_CONFIG), PERMISSION_WRITE)
                serverToClientCharacteristic?.addDescriptor(descriptorServerToClient)
                bluetoothGattService?.addCharacteristic(serverToClientCharacteristic)

                // Client2Server
                clientToServerCharacteristic = bleServiceCharacteristics.client2ServerCharacteristic
                bluetoothGattService?.addCharacteristic(clientToServerCharacteristic)

                // State
                stateCharacteristic = bleServiceCharacteristics.stateCharacteristic
                val descriptorState = BluetoothGattDescriptor(
                    UUID.fromString(CLIENT_CHARACTERISTIC_CONFIG), PERMISSION_WRITE)
                stateCharacteristic?.addDescriptor(descriptorState)
                bluetoothGattService?.addCharacteristic(stateCharacteristic)

                // Ident
                identCharacteristic = bleServiceCharacteristics.identCharacteristic
                bluetoothGattService?.addCharacteristic(identCharacteristic)
                identCharacteristic?.value = publicKeyHash

                Log.d(LOG_TAG, "Setup with Service: ${bluetoothGattService?.uuid}, " +
                        "serverToClientCharacteristic: ${serverToClientCharacteristic?.uuid} " +
                        "with Value: ${serverToClientCharacteristic?.value}, " +
                        "clientToServerCharacteristic: ${clientToServerCharacteristic?.uuid} " +
                        "with Value: ${clientToServerCharacteristic?.value}, " +
                        "stateCharacteristic: ${stateCharacteristic?.uuid} with Value: " +
                        "${stateCharacteristic?.value}, " +
                        "identCharacteristic: ${identCharacteristic?.uuid} with Value: " +
                        ByteUtils.bytesToHex(identCharacteristic?.value)
                )
            }
        } catch (ex: IllegalArgumentException) {
            Log.e(LOG_TAG, ex.message, ex)
        }
    }

    //region GattServiceInterface

    fun startServer() {
        getBluetoothGattServer().addService(bluetoothGattService)
        getPeripheralEventListener().onBLEEvent("Service UUID = ${bluetoothGattService?.uuid}",
            EventType.GATT_SERVICE_STARTED)
    }

    override fun onServiceAdded(status: Int, service: BluetoothGattService?) {

        if (bluetoothGattService == service) {
            if (BluetoothGatt.GATT_SUCCESS == status) {
                Log.d(javaClass.simpleName, "Service started.")
            } else {
                Log.d(javaClass.simpleName, "Service failed to start?: $status")
                getPeripheralEventListener().onBLEEvent("Service not successfully added", EventType.ERROR)
            }
        }
    }

    fun stopServer() {
        if (getBluetoothGattServer().getService(bluetoothGattService?.uuid) != null) {
            getBluetoothGattServer().removeService(bluetoothGattService)
        }
    }

    //region ITransportLayer
    fun setDelegate(executorEventListener: IExecutorEventListener) {
        this.executorEventListener = executorEventListener
    }

    fun close() {
        try {
            currentDevice.takeIf { it != null }.apply {
                getBluetoothGattServer().cancelConnection(currentDevice)
            }
            getBluetoothGattServer().clearServices()
            getBluetoothGattServer().close()
        } catch (ex: BluetoothException) {
            Log.e(LOG_TAG, ex.localizedMessage, ex)
        } catch (ex: NullPointerException) {
            Log.e(LOG_TAG, ex.localizedMessage, ex)
        }
        currentDevice = null
    }

    fun write(data: ByteArray, chunkSize: Int) {
        serverToClientCharacteristic?.let {
            dataNotifier = DataNotifier(data, chunkSize, getBluetoothGattServer(), it, getPeripheralEventListener(), currentDevice)
            val notifier = dataNotifier
            if (notifier != null)
                notifier.notifyInParts()
            else
                getPeripheralEventListener().onBLEEvent("dataNotifier is null", EventType.ERROR)
        }
    }
    //endregion

    fun writeToState(status: Byte) {
        Log.i(javaClass.simpleName, "Gatt server writeToState")
        stateCharacteristic?.value = byteArrayOf(status)
        val stateStatusWritten = getBluetoothGattServer().notifyCharacteristicChanged(currentDevice,stateCharacteristic,false)
        Log.i(javaClass.simpleName, "writeCharacteristic(${ByteUtils.bytesToHex(stateCharacteristic?.value)}): status = $status, Successfully written = $stateStatusWritten")
        if (!stateStatusWritten) {
            throw GattException("Could not write characteristic ${stateCharacteristic?.uuid} (value ${ByteUtils.bytesToHex(stateCharacteristic?.value)}.")
        }
    }

    //region BluetoothGattServerCallback


    @Synchronized
    override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
        Log.d(javaClass.simpleName, "onConnectionStateChange: ${deviceString(device)}, " +
                status + ", " + newState)
        //todo: check
        status.takeIf { it == BluetoothGatt.GATT_SUCCESS }?.apply {
            when(newState) {
                BluetoothGatt.STATE_CONNECTED -> {
                    if (currentDevice == null) {
                        currentDevice = device
                        getPeripheralEventListener().onBLEEvent("To device ${deviceString(device)}", EventType.GATT_CONNECTED)
                    }
                    else {
                        Log.i(javaClass.simpleName, "Already connected ${deviceString(device)}")
                    }
                }
                BluetoothGatt.STATE_DISCONNECTED -> {
                    getPeripheralEventListener().onBLEEvent("Disconnected from remote device",
                            EventType.GATT_DISCONNECTED)
                }
            }
        }
    }

    @Synchronized
    override fun onMtuChanged(device: BluetoothDevice?, mtu: Int) {
        this.maxSupportedMtu = mtu
    }

    @Synchronized
    override fun onCharacteristicReadRequest(device: BluetoothDevice, requestId: Int, offset: Int,
                                             characteristic: BluetoothGattCharacteristic
    ) {
        Log.i(javaClass.simpleName, "onCharacteristicReadRequest: ${characteristic.uuid}, " +
                "from ${deviceString(device)}, RequestId: $requestId, offset: $offset")

        when(characteristic.uuid) {
            identCharacteristic?.uuid -> {
                getBluetoothGattServer().sendResponse(device, requestId,
                    BluetoothGatt.GATT_SUCCESS, 0, publicKeyHash)
            }
        }
    }

    override fun onCharacteristicWriteRequest(device: BluetoothDevice, requestId: Int,
                                              characteristic: BluetoothGattCharacteristic,
                                              preparedWrite: Boolean, responseNeeded: Boolean,
                                              offset: Int, value: ByteArray?) {
        if (device.address != currentDevice?.address) {
            Log.i(javaClass.simpleName, "Unexpected device; returning error")
            getPeripheralEventListener().onBLEEvent("Unexpected device; returning error", EventType.ERROR)
            getBluetoothGattServer().sendResponse(device, requestId, BluetoothGatt.GATT_WRITE_NOT_PERMITTED, offset, null)
            return
        }
        this.requestId = requestId
        Log.i(javaClass.simpleName, "onCharacteristicWriteRequest: ${characteristic.uuid}, " +
                "from ${deviceString(device)}, RequestId: $requestId, Offset: $offset, " +
                "PreparedWrite: $preparedWrite and value: ${ByteUtils.bytesToHex(value)}")
        if (characteristic.uuid == bleServiceCharacteristics.stateUuid) {
            decodeState(value, requestId)
            return
        }
        if (characteristic.uuid != bleServiceCharacteristics.client2ServerUuid) {
            Log.i(javaClass.simpleName, "Unexpected characteristic.")
            return
        }
        Log.i(javaClass.simpleName, "using offset ${this.offset}")
        this.offset = offset
        if (value == null) {
            throw TransportLayerException("Received clientToServerCharacteristic value is null")
        }
        Log.i(javaClass.simpleName, "Current value of clientToServerCharacteristic: ${ByteUtils.bytesToHex(value)}")
        if (isDataPending(value)) {
            getPeripheralEventListener().onBLEEvent("Further data awaited", EventType.TRANSFER_IN_PROGRESS)
            dataStream.write(decodeData(value))
        } else {
            dataStream.write(decodeData(value))
            val receivedData = dataStream.toByteArray()
            Log.i("GattServer", "Reached last chunk of data after splitting into mtu-sized chunks: ${ByteUtils.bytesToHex(receivedData)}")

            Log.i("GattServer", "Delegated (to ITransportEventListener) data: ${ByteUtils.bytesToHex(receivedData)}")
            getDelegate().onReceive(receivedData)
            dataStream.reset()
        }
    }

    @Synchronized
    override fun onNotificationSent(device: BluetoothDevice?, status: Int) {
        Log.i(javaClass.simpleName, "onNotificationSent: device = $device, send status = $status")
        val notifier = dataNotifier
        if (notifier != null)
            notifier.notificationSentCalledBack()
        else
            getPeripheralEventListener().onBLEEvent("dataNotifier is null", EventType.ERROR)

    }

    private fun isDataPending(value: ByteArray): Boolean {
        return value[0] == DATA_PENDING
    }

    private fun decodeState(value: ByteArray?, requestId: Int) {
        if (value == null){
            throw GattException("Value of stateCharacteristic is null")
        }
        if (value.size != 1) {
            throw GattException("Invalid status")
        }
        if (value[0] == READY_FOR_TRANSMISSION) {
            getDelegate().onEvent(EventType.STATE_READY_FOR_TRANSMISSION.description, requestId)
            getPeripheralEventListener().onBLEEvent("Status updated", EventType.STATE_READY_FOR_TRANSMISSION)
        }
        if (value[0] == TERMINATE_TRANSMISSION) {
            getPeripheralEventListener().onBLEEvent("Status updated", EventType.GATT_DISCONNECTED)
        }
    }

    private fun getBluetoothGattServer(): BluetoothGattServer {
        return bluetoothGattServer ?: throw BluetoothException("bluetoothGattServer is null")
    }

    private fun getDelegate(): IExecutorEventListener {
        return executorEventListener ?: throw TransportLayerException("Delegate not set")
    }

    private fun getPeripheralEventListener() : BleEventListener {
        return bleEventListener ?: throw BluetoothException("Ble Event listener is null")
    }

    fun setReadyForNextFile(boolean: Boolean) {
        this.readyForNextFile = boolean
        
    }

    //endregion
}